contracts/staking/fuse/utils/StakingRewardsPerEpoch.sol

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract StakingRewardsPerEpoch is AccessControl, ReentrancyGuard, Pausable {
+	using SafeERC20 for IERC20;
+
+	struct StakeInfo {
+		uint256 rewardPerTokenPaid;
+		uint256 reward;
+		uint256 balance;
+		uint256 pendingStake;
+		uint256 indexOfLastEpochStaked;
+	}
+
+	event Staked(address indexed staker, uint256 amount);
+	event Withdrawn(address indexed staker, uint256 amount);
+	event RewardAdded(uint256 reward);
+	event RewardPaid(address indexed user, uint256 reward);
+	event RewardsDurationUpdated(uint256 newDuration);
+	event Recovered(address token, uint256 amount);
+
+	bytes32 public constant GUARDIAN_ROLE = keccak256("GUARDIAN_ROLE");
+	uint256 public constant PRECISION = 1e18;
+
+	IERC20 public rewardsToken;
+	IERC20 public stakingToken;
+
+	mapping(address => StakeInfo) public stakersInfo;
+
+	uint256 public lastEpochIndex;
+	uint256 public pendingStakes;
+	uint256[] public rewardsPerTokenAt;
+
+	uint256 internal _totalSupply;
+
+	modifier updateReward(address account) {
+		_updateReward(account);
+		_;
+	}
+
+	constructor(address _rewardsToken, address _stakingToken) {
+		_setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+		_setupRole(GUARDIAN_ROLE, msg.sender);
+		rewardsToken = IERC20(_rewardsToken);
+		stakingToken = IERC20(_stakingToken);
+	}
+
+	function totalSupply() external view returns (uint256) {
+		return _totalSupply;
+	}
+
+	function balanceOf(address account) external view returns (uint256) {
+		return stakersInfo[account].balance;
+	}
+
+	function stake(uint256 amount)
+		external
+		payable
+		nonReentrant
+		whenNotPaused
+		updateReward(msg.sender)
+	{
+		require(amount > 0, "Cannot stake 0");
+		_stake(msg.sender, amount);
+	}
+
+	function withdraw(uint256 _amount)
+		public
+		nonReentrant
+		updateReward(msg.sender)
+	{
+		require(_amount > 0, "Cannot withdraw 0");
+		_withdraw(msg.sender, _amount);
+	}
+
+	function getReward() public nonReentrant updateReward(msg.sender) {
+		_getReward(msg.sender);
+	}
+
+	function exit() external {
+		withdraw(stakersInfo[msg.sender].balance);
+		getReward();
+	}
+
+	function rewardPerToken() public view returns (uint256) {
+		if (_totalSupply == 0) {
+			return rewardPerTokenAt[0];
```

Comment:
isnt that just 0? why do we need a public rewardPerToken?
what happens if totalSupply was positive and then was reset back to 0?

Reply:
It's work in progress method. Quick answers: 

> isnt that just 0?
This is particular part that I am working on right now. So it will be altered.

> why do we need a public rewardPerToken?
Basically to let users know how much reward they can acquire per their one staked token.

> what happens if totalSupply was positive and then was reset back to 0?
The rewardPerToken must return previous value before the totalSupply was reset to zero.